extensions/legit/porcelain.lisp

```diff
@@ -109,554 +79,76 @@ Mercurial:
-     (values ".fslckout" :fossil))
-    (t
-     ;; Maybe do we need "fossil open" here.
-     (loop for file in (uiop:directory-files "./")
-           if (equal "fossil" (pathname-type file))
-             return (values file :fossil)))))
-
-(defun hg-project-p ()
-  "Return t if we find a .hg/ directory in the current directory (which should be the project root. Use `lem/porcelain:with-current-project`)."
-  (values (uiop:directory-exists-p ".hg")
-          :hg))
-
-(defvar *vcs-existence-order*
-  '(
-    git-project-p
-    fossil-project-p
-    hg-project-p
-    ))
-
-(defun vcs-project-p ()
-  "Is this project under a known version control system?
-  Return two values: the current directory (pathname), which is then considered the project root, and the VCS found (:git, :fossil or :hg)."
-  ;; This doesn't return the 2 values :(
-  ;; (or (fossil-project-p)
-  ;;     (git-project-p))
-  (loop for fn in *vcs-existence-order*
-        do (multiple-value-bind (root vcs)
-               (funcall fn)
-             (if root
-                 (return (values root vcs))))))
-
-(defun run-git (arglist)
-  "Run the git command with these options (list).
-  Return standard and error output as strings.
-
-  Don't signal an error if the process doesn't exit successfully
-  (but return the error message, so it can be displayed by the caller).
-  However, if uiop:run-program fails to run the command, the interactive debugger
-  is invoked (and shown correctly in Lem)."
-  (uiop:run-program (concatenate 'list
-                                 *git-base-arglist*
-                                 (uiop:ensure-list arglist))
-                    :output :string
-                    :error-output :string
-                    :ignore-error-status t))
-
-(defun run-fossil (arglist)
-  "Run the fossil command with these options.
-
-   arglist: a list of CLI options and arguments to append to the base fossil program.
-   See `*fossil-base-args*`."
-  (uiop:run-program (concatenate 'list
-                                 *fossil-base-args*
-                                 (uiop:ensure-list arglist))
-                    :output :string
-                    :error-output :string
-                    :ignore-error-status t))
-
-(defun run-hg (arglist)
-  "Run the mercurial command with these options (list).
-  Return standard and error output as strings.
-
-  For error handling strategy, see `run-git`."
-  (uiop:run-program (concatenate 'list
-                                 *hg-base-arglist*
-                                 (uiop:ensure-list arglist)
-                                 '("--pager" "never")   ;; no interactive pager.
-                                 )
-                    :output :string
-                    :error-output :string
-                    :ignore-error-status t))
-
-
 ;;;
 ;;; Getting changes.
 ;;;
+(defgeneric components (vcs)
+  (:documentation "Returns three values: TODO document what these are"))
```

Comment:
An inline "todo".

Reply:
Not sure what you mean. What exactly do you want?

Reply:
you wrote "TODO: document what these are" in your new docstring, so please write a proper docstring.